In [1]:
import pandas as pd
import numpy as np
#imported 2 libraries called the numpy and the pandas

This is the data being used for the peer graded assessment.
The dataset revolves around the compressive strength of various concrete samples, depending on the quantities of diverse components employed in their creation. These components encompass:

1) Cement

2) Blast Furnace Slag

3) Fly Ash

4) Water

5) Superplasticizer

6) Coarse Aggregate

7) Fine Aggregate

In [3]:
#Reading the data through the data dataframe 
concrete_data = pd.read_csv('concrete_data.csv')
concrete_data.head()  #printing the first five rows of the dataset.

Cement  Blast Furnace Slag  Fly Ash  Water  Superplasticizer  \
0   540.0                 0.0      0.0  162.0               2.5   
1   540.0                 0.0      0.0  162.0               2.5   
2   332.5               142.5      0.0  228.0               0.0   
3   332.5               142.5      0.0  228.0               0.0   
4   198.6               132.4      0.0  192.0               0.0   

   Coarse Aggregate  Fine Aggregate  Age  Strength  
0            1040.0           676.0   28     79.99  
1            1055.0           676.0   28     61.89  
2             932.0           594.0  270     40.27  
3             932.0           594.0  365     41.05  
4             978.4           825.5  360     44.30

So here we know that the first concrete sample has 540 cubic meter of cement, zero blast furnace slag, zero fly ash, 162 cubic meter of water,2.5 cubic meter of superplatsicizer, 1040 cubic meter of coarse aggregate, 676 cubic meter of fine aggregate, 28 is the age and 79.99 is the strength.

In [4]:
#Checking the total number of data points 
concrete_data.shape

(1030, 9)

In [5]:
#Checking data for any missing values 
concrete_data.describe()

Cement  Blast Furnace Slag      Fly Ash        Water  \
count  1030.000000         1030.000000  1030.000000  1030.000000   
mean    281.167864           73.895825    54.188350   181.567282   
std     104.506364           86.279342    63.997004    21.354219   
min     102.000000            0.000000     0.000000   121.800000   
25%     192.375000            0.000000     0.000000   164.900000   
50%     272.900000           22.000000     0.000000   185.000000   
75%     350.000000          142.950000   118.300000   192.000000   
max     540.000000          359.400000   200.100000   247.000000   

       Superplasticizer  Coarse Aggregate  Fine Aggregate          Age  \
count       1030.000000       1030.000000     1030.000000  1030.000000   
mean           6.204660        972.918932      773.580485    45.662136   
std            5.973841         77.753954       80.175980    63.169912   
min            0.000000        801.000000      594.000000     1.000000   
25%            0.000000        932.000000      730.950000     7.000000   
50%            6.400000        968.000000      779.500000    28.000000   
75%           10.200000       1029.400000      824.000000    56.000000   
max           32.200000       1145.000000      992.600000   365.000000   

          Strength  
count  1030.000000  
mean     35.817961  
std      16.705742  
min       2.330000  
25%      23.710000  
50%      34.445000  
75%      46.135000  
max      82.600000

In [6]:
# Removing all null data from the dataset 
concrete_data.isnull().sum()

Cement                0
Blast Furnace Slag    0
Fly Ash               0
Water                 0
Superplasticizer      0
Coarse Aggregate      0
Fine Aggregate        0
Age                   0
Strength              0
dtype: int64

As the data has no null values so now we can use this data to train our model.

For creating a Regression or a classification model we always need to divide the columns into two called the predictors and the target.

In [7]:
# Our target variable is Strength and others columns are the predictors
concrete_data_columns = concrete_data.columns
predictors = concrete_data[concrete_data_columns[concrete_data_columns != 'Strength']] # Every columns except Strength
target = concrete_data['Strength']  # Strength column

Printing the predictors and the targets just to check the output

In [10]:
target.head()  #As we have only 1 target variable so one column gets printed

0    79.99
1    61.89
2    40.27
3    41.05
4    44.30
Name: Strength, dtype: float64

In [11]:
predictors.head()

Cement  Blast Furnace Slag  Fly Ash  Water  Superplasticizer  \
0   540.0                 0.0      0.0  162.0               2.5   
1   540.0                 0.0      0.0  162.0               2.5   
2   332.5               142.5      0.0  228.0               0.0   
3   332.5               142.5      0.0  228.0               0.0   
4   198.6               132.4      0.0  192.0               0.0   

   Coarse Aggregate  Fine Aggregate  Age  
0            1040.0           676.0   28  
1            1055.0           676.0   28  
2             932.0           594.0  270  
3             932.0           594.0  365  
4             978.4           825.5  360

After printing the predictors and the target lets normalize the version of the data by subtracting the mean and diving by the standard deviation

In [41]:
predictors_norm = (predictors - predictors.mean()) / predictors.std()
predictors_norm.head()

Cement  Blast Furnace Slag   Fly Ash     Water  Superplasticizer  \
0  2.476712           -0.856472 -0.846733 -0.916319         -0.620147   
1  2.476712           -0.856472 -0.846733 -0.916319         -0.620147   
2  0.491187            0.795140 -0.846733  2.174405         -1.038638   
3  0.491187            0.795140 -0.846733  2.174405         -1.038638   
4 -0.790075            0.678079 -0.846733  0.488555         -1.038638   

   Coarse Aggregate  Fine Aggregate       Age  
0          0.862735       -1.217079 -0.279597  
1          1.055651       -1.217079 -0.279597  
2         -0.526262       -2.239829  3.551340  
3         -0.526262       -2.239829  5.055221  
4          0.070492        0.647569  4.976069

In [55]:
#Checking the number of predictors 
num_cols = predictors_norm.shape[1]
num_cols

8

Moving forward we import keras and start building our regression model

In [56]:
import keras     #Keras is always imported from the backend tensorflow

In [57]:
from keras.models import Sequential    #We are importing the sequential layer from keras which helps us to create a model layer by layer in a linear stack
from keras.layers import Dense    #A dense layer is imported from layers to form a fully connected neural network layer

Creating the regression model

In [58]:
def regression_model():    #Created a function to create a regression model.
    model = Sequential()
    model.add(Dense(10, activation='relu', input_shape=(num_cols,)))   #This contains 10 nodes with the relu activation function(1st hidden layer)
    model.add(Dense(1))  #Ouput layer

    model.compile(optimizer='adam', loss='mean_squared_error')  #Compiling the model with the optimizer adam and keeping the loss as MSE.
    return model

Now we want to split the data so we will have to import sklearn to train and test the data

In [59]:
from sklearn.model_selection import train_test_split

Splitting the data into 70 % training and 30% testing data

In [60]:
X_train, X_test, y_train, y_test = train_test_split(predictors_norm, target, test_size=0.3, random_state=42)

Now we have to build our model so we call the regression function here that we created above

In [61]:
#Calling our model here
model = regression_model()


In [62]:
#Training our model with epochs=50
epochs = 50
model.fit(X_train, y_train, epochs=epochs, verbose=1)

Epoch 1/50
721/721 [==============================] - 0s 306us/step - loss: 1586.7317
Epoch 2/50
721/721 [==============================] - 0s 101us/step - loss: 1566.6376
Epoch 3/50
721/721 [==============================] - 0s 87us/step - loss: 1546.9641
Epoch 4/50
721/721 [==============================] - 0s 105us/step - loss: 1527.1959
Epoch 5/50
721/721 [==============================] - 0s 85us/step - loss: 1507.6239
Epoch 6/50
721/721 [==============================] - 0s 84us/step - loss: 1487.6970
Epoch 7/50
721/721 [==============================] - 0s 78us/step - loss: 1467.7182
Epoch 8/50
721/721 [==============================] - 0s 61us/step - loss: 1447.1445
Epoch 9/50
721/721 [==============================] - 0s 74us/step - loss: 1426.3094
Epoch 10/50
721/721 [==============================] - 0s 58us/step - loss: 1404.9607
Epoch 11/50
721/721 [==============================] - 0s 58us/step - loss: 1382.8026
Epoch 12/50
721/721 [==============================] - 0s 59

Now we evaluate the metrics for the model formed 

In [63]:
loss_val = model.evaluate(X_test, y_test)
y_pred = model.predict(X_test)
loss_val

309/309 [==============================] - 0s 137us/step


379.75879034641105

As per the next step we need to compute the mean squared error

In [64]:
from sklearn.metrics import mean_squared_error

In [66]:
#Printing the mean and standard deviation of the test data 
mean_square_error = mean_squared_error(y_test, y_pred)
mean = np.mean(mean_square_error)
standard_deviation = np.std(mean_square_error)
print(mean, standard_deviation)

379.75878673098555 0.0


Creating a list of 50 mean squared errors.

In the code below it iterates through a specified number of training and evaluation cycles (total_mean_squared_errors). In each cycle, the dataset is split into training and testing sets, and a neural network model is trained for a fixed number of epochs. The mean squared error (MSE) is then calculated for the model's predictions on the test set, and both the MSE and the mean squared error between true and predicted values are recorded. After the specified number of cycles, the code calculates and prints the mean and standard deviation of the collected mean squared errors, providing insights into the model's overall performance and its variability across multiple runs.

In [67]:
total_mean_squared_errors = 50
#The total number of times the model will be trained and evaluated.
epochs = 50
mean_squared_errors = []
#An empty list to store the mean squared errors for each run.


for i in range(0, total_mean_squared_errors):
    X_train, X_test, y_train, y_test = train_test_split(predictors_norm, target, test_size=0.3, random_state=i)
    model.fit(X_train, y_train, epochs=epochs, verbose=0)
    MSE = model.evaluate(X_test, y_test, verbose=0)
    print("MSE "+str(i+1)+": "+str(MSE))
    y_pred = model.predict(X_test)
    mean_square_error = mean_squared_error(y_test, y_pred)
    mean_squared_errors.append(mean_square_error)

#Convert the list of mean squared errors to a NumPy array for easier calculation.
#Calculate the mean and standard deviation of the mean squared errors.
mean_squared_errors = np.array(mean_squared_errors)
mean = np.mean(mean_squared_errors)
standard_deviation = np.std(mean_squared_errors)

print('\n')
print("Printing the  mean and standard deviation of " +str(total_mean_squared_errors) + " mean squared errors with normalized data. Total number of epochs for each training is: " +str(epochs) + "\n")
print("Mean: "+str(mean))
print("Standard Deviation: "+str(standard_deviation))

MSE 1: 160.79323239619677
MSE 2: 146.63799179719103
MSE 3: 92.32815359171154
MSE 4: 67.80022983489299
MSE 5: 54.448322654156236
MSE 6: 52.78622558741893
MSE 7: 49.273305380614445
MSE 8: 37.825434687839746
MSE 9: 37.162871907058275
MSE 10: 36.44466685322882
MSE 11: 34.9958074254897
MSE 12: 33.05420955176492
MSE 13: 38.99397486467578
MSE 14: 40.86150532941602
MSE 15: 32.09869331680841
MSE 16: 26.668600495964963
MSE 17: 32.73275275369292
MSE 18: 31.748672108819957
MSE 19: 31.674569521907078
MSE 20: 31.5180233581938
MSE 21: 29.29480901428025
MSE 22: 32.18103780406964
MSE 23: 27.308880913990603
MSE 24: 29.10627881145786
MSE 25: 32.890351144241286
MSE 26: 34.6067228348124
MSE 27: 27.35724500390704
MSE 28: 29.75350922519721
MSE 29: 32.99266350385055
MSE 30: 30.672754176612038
MSE 31: 27.179041371762175
MSE 32: 27.053104122865548
MSE 33: 27.370280034333756
MSE 34: 29.909955799386726
MSE 35: 32.64390040524184
MSE 36: 35.29638431141677
MSE 37: 27.382841829343135
MSE 38: 30.517000871183033
MSE 39

Comparing the mean squared errors (MSE) between the two sets of results, one with normalized data and the other without, reveals notable differences in their magnitudes. In the case of the model trained without normalization (Step A), the MSE values range from approximately 45 to 67, with a mean of 52.87 and a standard deviation of 5.37. These values represent the model's performance on the original, non-normalized data.

On the other hand, for the model trained with normalized data (Step B), the MSE values span a wider range, from around 26 to 160. The mean MSE is notably lower at 39.44, but the standard deviation is substantially higher at 25.95. This suggests that while the normalized data helps achieve a lower average error, the model's performance exhibits greater variability across different runs compared to the non-normalized scenario.

In interpretation, the normalization process seems to have a positive impact on the average performance of the model, as indicated by the lower mean MSE. However, the higher standard deviation in the normalized case suggests that the model's performance can be more inconsistent across different datasets or runs. Further analysis and tuning may be necessary to understand and potentially address the increased variability in the normalized model's performance.






